tests/test_header.py

```diff
@@ -53,21 +82,84 @@ def test_linux_template_processing(
 
 @pytest.mark.parametrize("arch", ["x86_64", "arm64"])
 @pytest.mark.parametrize("check_path_spaces", [False, True])
-@pytest.mark.parametrize(
-    "script",
-    [
-        "checks_before_install.sh",
-        "prepare_installation.sh",
-        "run_installation.sh",
-        "update_path.sh",
-        "clean_cache.sh",
-        "run_user_script.sh",
-    ]
-)
-def test_osxpkg_template_processing(arch, check_path_spaces, script):
-    with open(os.path.join(OSX_DIR, script)) as f:
+@pytest.mark.parametrize("script", sorted(Path(OSX_DIR).glob("*.sh")))
+def test_osxpkg_scripts_template_processing(arch, check_path_spaces, script):
+    with script.open() as f:
         data = f.read()
     processed = preprocess(data, {"arch": arch, "check_path_spaces": check_path_spaces})
     assert "#if" not in processed
     assert "#else" not in processed
     assert "#endif" not in processed
+
+
+@pytest.mark.skipif(available_command("shellcheck") is False, reason="requires shellcheck")
+@pytest.mark.parametrize("arch", ["x86_64", "arm64"])
+@pytest.mark.parametrize("check_path_spaces", [False, True])
+@pytest.mark.parametrize("script", sorted(Path(OSX_DIR).glob("*.sh")))
+def test_osxpkg_scripts_shellcheck(arch, check_path_spaces, script):
+    with script.open() as f:
+        data = f.read()
+    processed = preprocess(data, {"arch": arch, "check_path_spaces": check_path_spaces})
+
+    findings, returncode = run_shellcheck(processed)
+    assert findings == []
+    assert returncode == 0
+
+
+@pytest.mark.skipif(available_command("shellcheck") is False, reason="requires shellcheck")
+@pytest.mark.parametrize("osx", [False, True])
+@pytest.mark.parametrize("direct_execute_post_install", [True])
+@pytest.mark.parametrize("direct_execute_pre_install", [True])
+@pytest.mark.parametrize("batch_mode", [True])
+@pytest.mark.parametrize("keep_pkgs", [True])
+@pytest.mark.parametrize("has_conda", [False, True])
+@pytest.mark.parametrize("has_license", [True])
+@pytest.mark.parametrize("initialize_conda", [True])
+@pytest.mark.parametrize("initialize_by_default", [True])
+@pytest.mark.parametrize("has_post_install", [True])
+@pytest.mark.parametrize("has_pre_install", [False])
+@pytest.mark.parametrize("arch", ["x86_64", "aarch64"])
+@pytest.mark.parametrize("check_path_spaces", [True])
```

Comment:
I know this is how we are checking other templated scripts but man is it getting out of hand 😂 Not a suggestion for now, but we need to come up with a different mechanism that doesn't suffer from combinatorial explosions so much. (Note to self!)

Reply:
yep ... leads to >20k test cases if a parameter options are given... executing shellcheck would take forever due subshell and command startup. reduced the options to have only 16 shellcheck runs.

we could stop templating the shellcripts and instead have some parameters defined at the beginning and use real if/else shell conditionals instead of templating if/else conditionals... that would result in one shellcheck run to check everything :) ... every .sh. installer would have the same header.sh attached, except of the variable at the beginning.